docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+Verona side, and a dummy allocator on the sandbox side that redirects all its
+calls to the Verona allocator. This is part of the sandbox library's
+functionality, which is passed along on the creation of the sandbox.
+
+The Verona allocator creates a heap for the sandbox and that's the memory that
+the sandbox code can use for everything. The sandbox code can also access that
+memory and it uses to keep the arguments and return values, so that the sandbox
+can see them.
+
+Because that heap is in a Verona sandbox region, it is safe to assume there are
+no race conditions introduced by the compiler onto the external library, due to
+the use of cowns to execute the foreign code.
+
+## Step-by-step example
+
+This is the description of a mock implementation of the child process sandbox
+running the C++ snippet in the beginning of this document. We'll go line by line
+and describe what the compiler and the generated code will do for each of them.
+
+### Compile time
+
+```c
+type StdVec = "c++::vector";
+```
+
+* Call the C++ driver (clang wrapper) to parse the include file (and all its
+  includes) and keep as a C++ AST.
+* Create a Verona class called `StdVec` with C++ built-in types defined as Verona
+  types (ex. `int` as `S32`, `size_t` as `U64`).
+* Expose an interface for the parser to query symbols (types, objects and
+  functions).
+
+```c
+reg = new StdVec;
+```
+
+* The parser recognises the region as a C++ sandbox.
+* Lower code to create a region with a sandbox allocator (RT calls).
+* Lower code to create the sandbox memory area (RT allocator calls).
+* Create a wrapper object for function dispatch and general sandbox utilities.
+* The region is bound to the variable `reg`.
+
+```c
+obj = new std::vector[int] in reg
+```
+
+* The parser recognises the region and tries to create a new object in the
+  sandbox region by using the C++ driver.
+* Assemble the foreign call signature (namespace, template args):
+  `std::vector<int> std::vector<int>::vector<int>()`
+* Instantiate the signature as a template specialisation in the AST, using
+  the query interface to validate the code and get an actual implementation.
+* If available, create the Verona equivalent inside the `StdVec` class and
+  Lower call to that function. Otherwise emit an error and stop processing.
+* The Verona signature would be something like
+  `create() : StdVec::std::vector[S32]`.
+* Generate the wrapper function from memory buffer to return value.
+* Add that function to the sandbox dispatcher's table and associate the index
+  with this particular specialisation, increment the index.
+* On the recently created `StdVec` Verona function, lower Verona code to call the
+  dispatcher with the index and the buffer and to extract the return value.
+* Bind the result of the Verona call to the variable `obj`.
+
+
+```c
+obj.push_back(42)
+```
+
+* The parser recognises the object is in a sanbdox region.
```

Comment:
The parser doesn't know anything.  The type inference pass should be sufficient to tell us that this is a foreign type in a specific kind of sandboxed region.  We don't need region analysis.  The main thing we get from region analysis is that we disallow a load of unsafe things.